src/server/game/Anticheat/AnticheatMgr.cpp

```diff
@@ -0,0 +1,453 @@
+    if (!timeDiff)
+        timeDiff = 1;
+
+    // this is the distance doable by the player in 1 sec, using the time done to move to this point.
+    uint32 clientSpeedRate = distance2D * 1000 / timeDiff;
+
+    // we did the (uint32) cast to accept a margin of tolerance
+    if (clientSpeedRate > speedRate)
+    {
+        TC_LOG_TRACE("entities.player.character", "AnticheatMgr:: Speed-Hack detected on player %s (%s)", player->GetName().c_str(), player->GetGUID().ToString().c_str());
+        return true;
+    }
+
+    return false;
+}
+
+bool FlyHackCheck::OnCheck(Player* player, MovementInfo const& /* movementInfo */, uint16 /*opcode = 0*/) const
+{
+    if (!player->m_movementInfo.HasMovementFlag(MOVEMENTFLAG_FLYING))
+        return false;
+
+    if (player->HasAuraType(SPELL_AURA_FLY) ||
+        player->HasAuraType(SPELL_AURA_MOD_INCREASE_MOUNTED_FLIGHT_SPEED) ||
+        player->HasAuraType(SPELL_AURA_MOD_INCREASE_FLIGHT_SPEED))
+        return false;
+
+    TC_LOG_TRACE("entities.player.character", "AnticheatMgr:: Fly-Hack detected on player %s (%s)", player->GetName().c_str(), player->GetGUID().ToString().c_str());
+    return true;
+}
+
+bool WalkOnWaterCheck::OnCheck(Player* player, MovementInfo const& /* movementInfo */, uint16 /*opcode = 0*/) const
+{
+    if (!player->m_movementInfo.HasMovementFlag(MOVEMENTFLAG_WATERWALKING))
+        return false;
+
+    // if we are a ghost we can walk on water
+    if (!player->IsAlive())
+        return false;
+
+    if (player->HasAuraType(SPELL_AURA_FEATHER_FALL) ||
+        player->HasAuraType(SPELL_AURA_SAFE_FALL) ||
+        player->HasAuraType(SPELL_AURA_WATER_WALK))
+        return false;
+
+    TC_LOG_TRACE("entities.player.character", "AnticheatMgr:: Walk on Water - Hack detected on player %s (%s)", player->GetName().c_str(), player->GetGUID().ToString().c_str());
+    return true;
+}
+
+bool JumpHackCheck::OnCheck(Player* player, MovementInfo const& /*movementInfo*/, uint16 opcode /*= 0*/) const
+{
+    AnticheatData* playerData = GetPlayerData(player->GetGUID());
+    if (!playerData)
+        return false;
+
+    if (playerData->GetLastOpcode() == MSG_MOVE_JUMP && opcode == MSG_MOVE_JUMP)
+    {
+        TC_LOG_TRACE("entities.player.character", "AnticheatMgr:: Jump-Hack detected on player %s (%s)", player->GetName().c_str(), player->GetGUID().ToString().c_str());
+        return true;
+    }
+
+    return false;
+}
+
+bool TeleportPlaneCheck::OnCheck(Player* player, MovementInfo const& movementInfo, uint16 /*opcode = 0*/) const
+{
+    if (player->m_movementInfo.pos.GetPositionZ() != 0 ||
+        movementInfo.pos.GetPositionZ() != 0)
+        return false;
+
+    if (movementInfo.HasMovementFlag(MOVEMENTFLAG_FALLING))
+        return false;
+
+    //DEAD_FALLING was deprecated
+    //if (player->getDeathState() == DEAD_FALLING)
+    //    return;
+    float x, y, z;
+    player->GetPosition(x, y, z);
+    float ground_Z = player->GetMap()->GetHeight(x, y, z);
```

Comment:
Hi everyone.

It has been long time since I wrote the original AC code. I can't really recall the idea behind that check.

I think that I could be trying not to divide by zero (in a trigonometric function?) in the GetHeight function, I don't know if that could really happen.
Another reason could be that I detected something in the  WoWEmuHacker's (for 3.3.5) behaviour.

Independently of the reasons above explained, I think that the check is not correct. GetPositionZ is a float number, so it must be checked against an epsilon.

About the performance concerns, at that moment I did not understand the TC threading model. Now I understand that anticheat checks are made in the same thread that is processing packets, so it is not a nice idea to perform heavy computations. 

I think that the options available to mitigate this drawback, if the anticheat checks are made in the same thread, are the followings:

Randomize this check, I mean it could happen with certain probability. If you are detected positive, your probability increases otherwise decreases.
Inspecting the GetHeight() function I noticed that there are two metrics used: maps data and vmaps data. We could consider just using maps data, in case it were faster than vmap computations (I don't really know about the peformance costs of each).

Finally, we could consider doing all this checks in a separated thread, in case that map or vmap height were thread safe. @Shauren Could you enlighten us about this? 

If I am not wrong vmap height is thread-safe unless you set isHeightCalcEnabled as false, but it is just set at startup.